pynicotine/shares.py

```diff
@@ -485,55 +507,61 @@ def init_shares(self):
 
         self.rescan_shares(init=True, rescan=rescan_startup)
 
-    def real2virtual(self, path):
-        return self.real2virtual_cls(path, self.config)
-
-    @classmethod
-    def real2virtual_cls(cls, path, config):
+    @staticmethod
+    def _virtual2real(shares, virtual_path):
 
-        path = path.replace('/', '\\')
+        virtual_path = virtual_path.replace('/', os.sep).replace('\\', os.sep)
+        virtual_name = virtual_path.split(os.sep, 1)[0]
 
-        for virtual, real, *_unused in cls._virtualmapping(config):
+        for virtual_name_shared, real_folder_path, *_unused in shares:
             # Remove slashes from share name to avoid path conflicts
-            virtual = virtual.replace('/', '_').replace('\\', '_')
+            virtual_name_shared = virtual_name_shared.replace('/', '_').replace('\\', '_')
+
+            if virtual_name_shared != virtual_name:
+                continue
 
-            real = real.replace('/', '\\')
-            if path == real:
-                return virtual
+            if virtual_name == virtual_path:
+                return real_folder_path
 
-            # Use rstrip to remove trailing separator from root directories
-            real = real.rstrip('\\') + '\\'
+            if not virtual_path.startswith(virtual_name + os.sep):
+                continue
 
-            if path.startswith(real):
-                virtualpath = virtual + '\\' + path[len(real):]
-                return virtualpath
+            real_path = real_folder_path + virtual_path[len(virtual_name):]
 
-        return "__INTERNAL_ERROR__" + path
+            if real_path != os.path.abspath(real_path):
+                # Protect against directory traversal
+                return None
 
-    def virtual2real(self, path):
+            folder_path, _sep, basename = virtual_path.rpartition(os.sep)
 
-        path = path.replace('/', os.sep).replace('\\', os.sep)
+            if Scanner.is_hidden(folder_path, basename):
+                return None
 
-        for virtual, real, *_unused in self._virtualmapping(self.config):
-            # Remove slashes from share name to avoid path conflicts
-            virtual = virtual.replace('/', '_').replace('\\', '_')
+            return real_path
 
-            if path == virtual:
-                return real
+        return None
 
-            if path.startswith(virtual + os.sep):
-                realpath = real + path[len(virtual):]
-                return realpath
+    def virtual2real(self, virtual_path, user=None):
```

Comment:
This method now contains code that seems unrelated to the name of this function.